Allow fontForCombiningCharacterSequence() and canRenderCombiningCharacterSequence() to be able to accept 8-bit characters

litherum opened the pull request:
#### 67c98f07295d7d27c34f307816941ec726a96c73
<pre>
Allow fontForCombiningCharacterSequence() and canRenderCombiningCharacterSequence() to be able to accept 8-bit characters
<a href="https://bugs.webkit.org/show_bug.cgi?id=260753">https://bugs.webkit.org/show_bug.cgi?id=260753</a>
rdar://114478724

Reviewed by Cameron McCormack.

If we&apos;re going to make the fast text codepath powerful enough to handle complex
text, that means it&apos;s going to have to call fontForCombiningCharacterSequence()
and canRenderCombiningCharacterSequence(). The fast text codepath is templated
to accept either 8-bit characters or 16-bit characters, which means anything
it calls must be able to do so as well.

There&apos;s no behavior change with just this patch - these functions are still only
being called with 16-bit characters. This patch just changes the type of them
so they could be called with 8-bit characters in the future. (So, there are no
tests in this patch.)

* Source/WebCore/platform/graphics/ComplexTextController.cpp:
(WebCore::ComplexTextController::collectComplexTextRuns):
* Source/WebCore/platform/graphics/Font.cpp:
(WebCore::Font::canRenderCombiningCharacterSequence const):
* Source/WebCore/platform/graphics/Font.h:
* Source/WebCore/platform/graphics/FontCascade.cpp:
(WebCore::FontCascade::fontForCombiningCharacterSequence const):
* Source/WebCore/platform/graphics/FontCascade.h:
* Source/WebCore/platform/graphics/coretext/FontCascadeCoreText.cpp:
(WebCore::FontCascade::fontForCombiningCharacterSequence const):

Canonical link: <a href="https://commits.webkit.org/267317@main">https://commits.webkit.org/267317@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4803eade50249d68445dc5e0ae3b55cfb9081ef6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/16276 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/16595 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/17013 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/18044 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15265 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/16464 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/19672 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/16707 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/17648 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/16470 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/16912 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/13928 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/18810 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/14156 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/14735 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/21552 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15144 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/14900 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/18163 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/15491 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/13130 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/30/builds/14712 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3889 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/19078 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/15318 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebCore/platform/graphics/FontCascade.cpp:
```diff
@@ -1536,11 +1536,10 @@ int FontCascade::offsetForPositionForComplexText(const TextRun& run, float x, bo
-    size_t baseCharacterLength = 0;
-    U16_NEXT(characters, baseCharacterLength, length, baseCharacter);
+    ASSERT(stringView.length() > 0);
+    UChar32 baseCharacte = *stringView.codePoints().begin();
```
```suggestion
    UChar32 baseCharacter = *stringView.codePoints().begin();
```

litherum commented on Source/WebCore/platform/graphics/FontCascade.cpp:
```diff
@@ -1536,11 +1536,10 @@ int FontCascade::offsetForPositionForComplexText(const TextRun& run, float x, bo
-    size_t baseCharacterLength = 0;
-    U16_NEXT(characters, baseCharacterLength, length, baseCharacter);
+    ASSERT(stringView.length() > 0);
+    UChar32 baseCharacte = *stringView.codePoints().begin();
```
typo!

litherum replied:
Maybe this is dead code? 🤔

litherum committed:
Allow fontForCombiningCharacterSequence() and canRenderCombiningCharacterSequence() to be able to accept 8-bit characters
https://bugs.webkit.org/show_bug.cgi?id=260753
rdar://114478724

Reviewed by Cameron McCormack.

If we're going to make the fast text codepath powerful enough to handle complex
text, that means it's going to have to call fontForCombiningCharacterSequence()
and canRenderCombiningCharacterSequence(). The fast text codepath is templated
to accept either 8-bit characters or 16-bit characters, which means anything
it calls must be able to do so as well.

There's no behavior change with just this patch - these functions are still only
being called with 16-bit characters. This patch just changes the type of them
so they could be called with 8-bit characters in the future. (So, there are no
tests in this patch.)

* Source/WebCore/platform/graphics/ComplexTextController.cpp:
(WebCore::ComplexTextController::collectComplexTextRuns):
* Source/WebCore/platform/graphics/Font.cpp:
(WebCore::Font::canRenderCombiningCharacterSequence const):
* Source/WebCore/platform/graphics/Font.h:
* Source/WebCore/platform/graphics/FontCascade.cpp:
(WebCore::FontCascade::fontForCombiningCharacterSequence const):
* Source/WebCore/platform/graphics/FontCascade.h:
* Source/WebCore/platform/graphics/coretext/FontCascadeCoreText.cpp:
(WebCore::FontCascade::fontForCombiningCharacterSequence const):

Canonical link: https://commits.webkit.org/267317@main
